app/models/student.rb

```diff
@@ -156,8 +154,23 @@ def submission_for(code_review)
     submissions.find { |submission| submission.code_review_id == code_review.id }
   end
 
+  def pairs(course=nil)
+    selected_attendance_records = course ? attendance_records.where("date between ? and ?", course.try(:start_date), course.try(:end_date)) : attendance_records
+    (selected_attendance_records.map {|s| s.pairings.pluck(:pair_id)}.flatten).compact.sort
+  end
+
   def pairs_on_day(day)
-    Student.where(id: attendance_record_on_day(day).try(:pair_ids))
+    Student.where(id: attendance_record_on_day(day).try(:pairings).try('pluck', 'pair_id'))
+  end
+
+  def inverse_pairs_on_day(day)
+    Student.where(id: AttendanceRecord.where(date: day).joins(:pairings).where(pairings: { pair_id: id }).pluck(:student_id))
+  end
+
+  def pairs_without_feedback_today
+    today = Time.zone.now.to_date
+    all_pairs = pairs_on_day(today).or(inverse_pairs_on_day(today))
+    all_pairs.select { |pair| !PairFeedback.today.find_by(student: self, pair: pair)}
```

Comment:
Can you change these two lines to use SQL/AR? It will be more efficient.

Reply:
Yes likely! I've actually just begun to change over in a few places from using select to trying to use joins etc to be as efficient as possible.

Reply:
Sweet!

Reply:
I was able to replace this with an AR query, but it feels a bit wordy. I imagine there might be a cleaner way. I mean obviously I could combine all of this into one line, but that doesn't seem like it would move things in the direction of being clearer.

```
def pairs_without_feedback_today
  today = Time.zone.now.to_date
  all_pairs = pairs_on_day(Time.zone.now.to_date).or(inverse_pairs_on_day(Time.zone.now.to_date))
  existing_feedback = PairFeedback.today.where(student: self, pair: all_pairs)
  all_pairs.where.not(id: existing_feedback.pluck(:pair_id))
end
```

Reply:
Actually scratch that. I realized that I no longer need to consider unreciprocated pairs for purposes of PairFeedback, because presumably by the end of the day any unreciprocated pairings should be fixed. And I realized I could clean this up to just:

```
  def pairs_without_feedback_today
    pairs_on_day(Time.zone.now.to_date).where.not(id: evaluations_of_pairs.today.pluck(:pair_id))
  end
```